waku/v2/discv5/discover.go

```diff
@@ -0,0 +1,357 @@
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/status-im/go-waku/waku/v2/utils"
+)
+
+var log = logging.Logger("waku_discv5")
+
+type DiscoveryV5 struct {
+	discovery.Discovery
+
+	params    *discV5Parameters
+	host      host.Host
+	config    discover.Config
+	udpAddr   *net.UDPAddr
+	listener  *discover.UDPv5
+	localnode *enode.LocalNode
+
+	peerCache peerCache
+}
+
+type peerCache struct {
+	sync.RWMutex
+	recs map[peer.ID]peerRecord
+	rng  *rand.Rand
+}
+
+type peerRecord struct {
+	expire int64
+	peer   peer.AddrInfo
+}
+
+type discV5Parameters struct {
+	bootnodes        []*enode.Node
+	advertiseAddress *net.IP
+	udpPort          int
+}
+
+const WakuENRField = "waku2"
+
+// WakuEnrBitfield is a8-bit flag field to indicate Waku capabilities. Only the 4 LSBs are currently defined according to RFC31 (https://rfc.vac.dev/spec/31/).
+type WakuEnrBitfield = uint8
+
+type DiscoveryV5Option func(*discV5Parameters)
+
+func WithBootnodes(bootnodes []*enode.Node) DiscoveryV5Option {
+	return func(params *discV5Parameters) {
+		params.bootnodes = bootnodes
+	}
+}
+
+func WithAdvertiseAddress(advertiseAddr net.IP) DiscoveryV5Option {
+	return func(params *discV5Parameters) {
+		params.advertiseAddress = &advertiseAddr
+	}
+}
+
+func WithUDPPort(port int) DiscoveryV5Option {
+	return func(params *discV5Parameters) {
+		params.udpPort = port
+	}
+}
+
+func DefaultOptions() []DiscoveryV5Option {
+	return []DiscoveryV5Option{
+		WithUDPPort(9000),
+	}
+}
+
+func NewWakuEnrBitfield(lightpush, filter, store, relay bool) WakuEnrBitfield {
+	var v uint8 = 0
+
+	if lightpush {
+		v |= (1 << 3)
+	}
+
+	if filter {
+		v |= (1 << 2)
+	}
+
+	if store {
+		v |= (1 << 1)
+	}
+
+	if relay {
+		v |= (1 << 0)
+	}
+
+	return v
+}
+
+func NewDiscoveryV5(host host.Host, ipAddr net.IP, tcpPort int, priv *ecdsa.PrivateKey, wakuFlags WakuEnrBitfield, opts ...DiscoveryV5Option) (*DiscoveryV5, error) {
+	params := new(discV5Parameters)
+	optList := DefaultOptions()
+	optList = append(optList, opts...)
+	for _, opt := range optList {
+		opt(params)
+	}
+
+	localnode, err := newLocalnode(priv, ipAddr, params.udpPort, tcpPort, wakuFlags, params.advertiseAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	discV5 := new(DiscoveryV5)
+	discV5.host = host
+	discV5.params = params
+	discV5.peerCache.rng = rand.New(rand.NewSource(rand.Int63()))
+	discV5.peerCache.recs = make(map[peer.ID]peerRecord)
+	discV5.localnode = localnode
+	discV5.config = discover.Config{
+		PrivateKey: priv,
+		Bootnodes:  params.bootnodes,
+	}
+	discV5.udpAddr = &net.UDPAddr{
+		IP:   ipAddr,
+		Port: params.udpPort,
+	}
+
+	return discV5, nil
+}
+
+func newLocalnode(priv *ecdsa.PrivateKey, ipAddr net.IP, udpPort int, tcpPort int, wakuFlags WakuEnrBitfield, advertiseAddr *net.IP) (*enode.LocalNode, error) {
+	db, err := enode.OpenDB("")
```

Comment:
why the empty string?

Reply:
An empty string means that a in-memory database will be used for storing the enodes. It's not necessary to persist the enode information since these will be converted to multiaddresses and we already have persistent peer storage for those